langchain4j/src/main/java/dev/langchain4j/service/AiServices.java

```diff
@@ -280,22 +278,141 @@ public AiServices<T> tools(Object... objectsWithTools) {
      * @see Tool
      */
     public AiServices<T> tools(List<Object> objectsWithTools) {
-        context.toolSpecifications = new ArrayList<>();
-        context.toolExecutors = new HashMap<>();
-
         for (Object objectWithTool : objectsWithTools) {
             for (Method method : objectWithTool.getClass().getDeclaredMethods()) {
-                if (method.isAnnotationPresent(Tool.class)) {
-                    ToolSpecification toolSpecification = toolSpecificationFrom(method);
-                    context.toolSpecifications.add(toolSpecification);
-                    context.toolExecutors.put(toolSpecification.name(), new ToolExecutor(objectWithTool, method));
-                }
+                if (!method.isAnnotationPresent(Tool.class)) continue;
+                final ToolExecutor toolExecutor = new ToolExecutor(objectWithTool, method);
+                final ToolSpecification toolSpecification = toolExecutor.toolSpecification();
+
+                context.addToolSpecification(toolSpecification);
+                context.putToolExecutor(toolSpecification.name(), toolExecutor);
             }
         }
 
         return this;
     }
 
+    public <TOOL, U1, R> AiServices<T> toolFunction(TOOL tool, ToolFunction<TOOL, U1, R> toolFunction, String description) {
```

Comment:
could you please show an example of how this will be used?

Reply:
see `AiServicesIT#should_execute_tool_function_then_answer`

https://github.com/BOFA1ex/langchain4j/blob/feat_tool_function/langchain4j/src/test/java/dev/langchain4j/service/AiServicesIT.java#L739-L745

Reply:
It could be more flexiable to reuse exist tool.
```java
static class PersonMapper {
     Person selectByNameAndBirthRange(String name, long birthL, long birthR);
}

.toolFunction(personMapper, PersonMapper::selectByNameAndBirthRange, ToolParamters.builder()
    .addParameter("name", STRING, JsonSchemaProperty.description("person name"))
    .addParameter("birthL", NUMBER, JsonSchemaProperty.description("person birth range of start"))
    .addParameter("birthR", NUMBER, JsonSchemaProperty.description("person birth range of end"))
    .build();
```

Reply:
if u wanna use the other dep tool, but u coun't mark with `@Tool`. it's work well.
```java
.toolFunction(apache.xxxUtils::convertPartSpecTox, ToolParamters.builder()
    .addParameter("arg0", STRING, JsonSchemaProperty.description(""))
    .addParameter("arg1", NUMBER, JsonSchemaProperty.description(""))
    .build();
```

Reply:
@BOFA1ex you can always wrap any code in the `@Tool`:

```java
@Tool
public String convertPartSpecTox(@P("description of arg0") String arg0, @P("description of arg1") Integer arg1) {
    return convertPartSpecTox(arg0, arg1);
}
```